bootloader/src/old_machine_bootloader_windows.c

```diff
@@ -200,76 +243,109 @@ char* getOsVersion(){
     }
 }
 
-int ping_island(int argc, char * argv[])
-{
+int ping_island(int argc, char * argv[]) {
     // Get all machine IP's
     int addrCount = 0;
     char* hostname;
     char** IPs = getIpAddresses(&addrCount, &hostname);
-    printf("hostname: %s\n", hostname);
-    printf("Addr Count: %d", addrCount);
-    char* IPstring = concatenate(addrCount, IPs, "\", \"");
-    printf("Concatenated ips: %s\n", IPstring);
+    char* IPstring;
+    if (IPs != NULL) {
+        IPstring = concatenate(addrCount, IPs, "\", \"");
+    } else {
+        IPstring = '\0';
+    }
 
     char* windowsVersion = getOsVersion();
     printf("Windows version: %s\n", windowsVersion);
 
     // Find which argument is tunnel flag
-    int i, tunnel_i=0, server_i=0;
-    for(i=1;i<argc;i++)
-    {
-        if(strcmp(argv[i], "-t") == 0 || strcmp(argv[i], "--tunnel") == 0){
-            tunnel_i = i+1;
-        } else if(strcmp(argv[i], "-s") == 0 || strcmp(argv[i], "--server") == 0){
-            server_i = i+1;
-        }
+    int tunnel_i, server_i;
+    int parse_error = parseFlags(argc, argv, &server_i, &tunnel_i);
+    if (parse_error) {
+        error("Flag parse failed\n");
     }
 
     // Form request struct
     struct requestData reqData;
     reqData.osVersion = windowsVersion;
     reqData.hostname = hostname;
     reqData.IPstring = IPstring;
-	reqData.tunnel = NULL;
-	reqData.glibcVersion = NULL;
+    reqData.tunnel = NULL;
+    reqData.glibcVersion = NULL;
+
+
+    BOOL requiredDllPresent = TRUE;
+    // If running on windows 7 monkey will crash if system is not updated
+    if (!strcmp(windowsVersion, WINDOWS7SP1) || !strcmp(windowsVersion, WINDOWS7)) {
+        TCHAR windir[MAX_PATH];
+        GetWindowsDirectory(windir, MAX_PATH);
```

Comment:
Why not use `GetSystemDirectory` and then you can avoid appending system32, which might not be accurate in all systems :)

Reply:
Also always check return value

Reply:
I didn't know that it's not always `windows/system32` 😮 . I'll add the fix